### ISSUE TITLE
BlendMode.ADD not working in Container with IdentityFilter

### ISSUE THREAD

**Issue by zhanghai** [korlibs/korge#800]
I tried to find the answer in Korge code but couldn't, so I'm posting here in the hope that you may know how to properly do it in Korge.

Basically I have a scenario where there are two overlapping 50% transparent white objects on an opaque black background - if we follow the painter's algorithm, it would give us a greyish result because the lower 50% transparent whtie is first blended over the opaque black background to produce grey, then the higher 50% transparent white is blended over this grey, producing a lighter grey.

However, in this case I would like to blend the two 50% transparent whtie objects first, to produce an opaque white, and then blend that white onto black so that it stays white. It seems to me I need to somehow render the two objects together to a "layer" first, but I wasn't able to find something similar for it in Korge. Do you know if there's anything that allows doing this in Korge?

(Note that the actual scenario is two similar but different overlapping images (bitmaps) crossfading together over an opaque background - so considering the animation, an efficient solution would be preferred.)

**Comment by soywiz** [comment]
If you want to treat a whole container as a whole for alpha blending, you can apply a filter to it.

```
container.filter = IdentityFilter
```

Please check this section in the documentation: https://docs.korge.org/korge/reference/filters/#identityfilter

**Comment by zhanghai** [comment]
That sounds pretty nice, but when I tried it didn't work. I've created an MWE based on the Hello World project and whether filter is specified doesn't seem to affect the outcome, which still gets semi-transparent during animation:

```kotlin
suspend fun main() = Korge(width = 1024, height = 512, bgcolor = Colors["#2b2b2b"]) {
    val image1 = image(resourcesVfs["korge.png"].readBitmap()) {
        anchor(.5, .5)
        position(256, 256)
    }
    val image2 = image(resourcesVfs["korge.png"].readBitmap()) {
        anchor(.5, .5)
        position(256, 256)
        alpha(0)
    }
    val container = container {
        filter = IdentityFilter
    }
	val image3 = container.image(resourcesVfs["korge.png"].readBitmap()) {
		anchor(.5, .5)
		position(768, 256)
	}
	val image4 = container.image(resourcesVfs["korge.png"].readBitmap()) {
		anchor(.5, .5)
		position(768, 256)
        alpha(0)
	}

    while (true) {
        tween(
            image1::alpha[1.0, 0.0], image2::alpha[0.0, 1.0], image3::alpha[1.0, 0.0],
            image4::alpha[0.0, 1.0], time = 1000.milliseconds, easing = Easing.LINEAR
        )
    }
}
```

**Comment by zhanghai** [comment]
I realized I also need `BlendMode.ADD` - however, the result image is still blinking, though in the brighter direction:

```kotlin
    val container = container {
        blendMode = BlendMode.ADD
        filter = IdentityFilter
    }
```

[recording2.webm](https://user-images.githubusercontent.com/4469895/178390968-e31c1205-a84d-4d97-a817-ad5950dd4fe7.webm)

It seems to me given that ADD makes `a = aA + aB`, `c = aA * cA + aB * cB`, and in my case `aA + aB == 1` and `cA == cB`, my image should stay the same during the animation, so I'm still confused...

**Comment by zhanghai** [comment]
It appears that the brightening effect in my last comment is the result of using `ADD` for composoting the container itself onto the grey background, and if I make the background black, the two becomes the same again.

I still didn't figure out a way to make the two images in the `IdentityFilter` container to use `ADD` - if I just set it on the two images, the result remained the same as the default `OVER`:

```
    val container = container {
        filter = IdentityFilter
    }
    val image3 = container.image(resourcesVfs["korge.png"].readBitmap()) {
        blendMode = BlendMode.ADD
        anchor(0, 0)
        position(512, 0)
    }
    val image4 = container.image(resourcesVfs["korge.png"].readBitmap()) {
        blendMode = BlendMode.ADD
        anchor(0, 0)
        position(512, 0)
        alpha(0)
    }
```

**Comment by zhanghai** [comment]
Okay I think I found out the issue here: `BlendMode.ADD` isn't working in a `Container` with an `IdentityFilter`:

With the following code, the issue is evident now:

```kotlin
suspend fun main() = Korge(width = 1024, height = 1024) {
    val noFilterContainer = container()
    val identityFilterContainer = container {
        filter = IdentityFilter
    }
    val bitmap = resourcesVfs["korge.png"].readBitmap()
    fun Container.plusLighterImage(x: Int, y: Int, alpha: Double) = image(bitmap) {
        blendMode = BlendMode.ADD
        anchor(0, 0)
        position(x, y)
        alpha(alpha)
    }
    val image1 = noFilterContainer.plusLighterImage(0, 0, 1.0)
    val image2 = noFilterContainer.plusLighterImage(0, 0, 0.0)
    val image3 = identityFilterContainer.plusLighterImage(512, 0, 1.0)
    val image4 = identityFilterContainer.plusLighterImage(512, 0, 0.0)
    val image5 = noFilterContainer.plusLighterImage(0, 512, 0.5)
    val image6 = noFilterContainer.plusLighterImage(0, 512, 0.5)
    val image7 = identityFilterContainer.plusLighterImage(512, 512, 0.5)
    val image8 = identityFilterContainer.plusLighterImage(512, 512, 0.5)

    while (true) {
        tween(
            image1::alpha[1.0, 0.0], image2::alpha[0.0, 1.0], image3::alpha[1.0, 0.0],
            image4::alpha[0.0, 1.0], time = 1000.milliseconds, easing = Easing.LINEAR
        )
    }
}
```

[record3.webm](https://user-images.githubusercontent.com/4469895/178409591-4fc518a3-0919-4f9d-8658-b7a94f7111db.webm)

**Comment by zhanghai** [comment]
This is kind of a blocker for myself - do you think this is something that can be fixed in the short term, and is there any way I can work around locally for this?

**Comment by soywiz** [comment]
I would say this is related to filters not using the right blend mode and views should use the normal blend mode if there is a filter:

https://github.com/korlibs/korge/blob/702222a1d49383b2d4e7871fb1a72c25ac47a3bf/korge/src/commonMain/kotlin/com/soywiz/korge/view/filter/ViewFilter.kt#L97
I guess this should be `blendMode` -> `renderBlendMode`. And also it should somehow disable the custom blendMode for the original view.

Not sure if there is a quick workaround. You can try to asume it works properly, and then get fixed at some point. I will try to fix it before 3.0, and I plan to release it by the end of the month.

**Comment by zhanghai** [comment]
The MWE I provided is blending the container onto a black background so whether the container itself is using BlendMode.ADD or BlendMode.NORMAL will produce the same result. This was done to allow the no-filter example to showcase the right visual result.

However, in a real world scenario, the container would be blending onto arbitrary background so it should only use `BlendMode.NORMAL` for itself, while it's children should use `BlendMode.ADD` while rendering to the container's texture/canvas/cache (not sure what it's called in Jorge).

Does this align with your described fix? I'm asking because your fix seems to suggest I'm setting `BlendMode.ADD` on the container, but in fact I'm only setting it for the children of the container, but not the container itself.

**Comment by soywiz** [comment]
Yeah, that scenario should work then. Maybe the problem is https://github.com/korlibs/korge/issues/506

**Comment by zhanghai** [comment]
Could you elaborate more on `filters not using the right blend mode` here?

I tried to do some debugging, but I was able to confirm that (if I didn't get anything wrong) `BlendMode.ADD` is correctly passed down into `AGList.setBlendingState`: https://github.com/korlibs/korge/blob/d78da683e36096597bb53a1bf065e813f97dfe79/korgw/src/commonMain/kotlin/com/soywiz/korag/AGListExt.kt#L20-L26

Specifically, by only rendering `image3` and `image4` and setting `identityContainer.blendMode = BlendMode.ADD` as well in the previous example, I could confirm that all values passed down to `blendFunction()` are always the values for `BlendMode.ADD`. The same for `BatchBuilder2D.setStateFast()`. However, the rendered result still behaves exactly the same as not setting `BlendMode.ADD` for the two animating images,

(For the original case, I was able to confirm that when drawing the image to the container texture `BlendMode.ADD` is correctly passed to `AGList`, while when using & drawing the texture `BlendMode.NORMAL` is correctly passed to `AGList` as well.)

So the only conclusion I could draw from this is that `BlendMode` is correctly passed down to the GL implementation, but the implemetation ignored it when rendering to a texture. What do you think?

Also I don't see how this is related to #506 about premultiplied alpha - It seems to me in the worst case of that issue, the rendering result could have artifacts, but I don't see a way to make it look exactly the same as another blend function (`ADD` v.s. `NORMAL`) just due to premultiplied alpha.

**Comment by zhanghai** [comment]
Okay I figured out what went wrong - it is indeed related to premultiplied alpha (#506).

After I changed the background to colors like white or red, I was able to confirm the supposedly-white area rendered to the texture was actually in an opque 75% grey instead of a white with transparency. And my suspected root cause is that when rendering to texture, the texture is in premultiplied format, but the images are not.

As a result, the outcome stays unaffected when only one image is renderred onto an empty transparent texture (since dest alpha is 0), but becomes wrong when a second image is rendered onto the texture where it's no longer fully transparent - the dest (texture) colors will get multiplied by a non-zero dest alpha twice and appear darker than it should have been. And this is the source of the grey I was seeing.

To verify this (and work around in a very ugly way), I was able to take an unused enum from `BlendMode` and use Java reflection to modify it to use `ONE` for `dstRgb`. Using this modified `BlendMode`, I was finally able to get the correct plus-lighter blending within the container while maintaining the normal blend mode when blending the container onto background.

```kotlin
    fun Container.plusLighterImage(x: Int, y: Int, alpha: Double) = image(bitmap) {
        blendMode = hackAddBlendMode
        anchor(0, 0)
        position(x, y)
        alpha(alpha)
    }
...
expect val hackAddBlendMode: BlendMode
```

```kotlin
actual val hackAddBlendMode: BlendMode
    get() {
        val factors = BlendMode::class.java.getDeclaredField("factors")
        factors.isAccessible = true
        factors.set(BlendMode.ALPHA, AG.Blending(
            AG.BlendFactor.SOURCE_ALPHA, AG.BlendFactor.ONE,
            AG.BlendFactor.ONE, AG.BlendFactor.ONE
        ))
        return BlendMode.ALPHA
    }
```
